src/Configuration/Programmer/Programmer_cmd.cpp

```diff
@@ -19,57 +19,145 @@ You should have received a copy of the GNU General Public License
+    CFG_POST_MSG("desc      : %s", cable.description.c_str());
+    CFG_POST_MSG("bus       : %d", cable.busAddr);
+    CFG_POST_MSG("port      : %d", cable.portAddr);
+    CFG_POST_MSG("device    : %d", cable.deviceAddr);
+    CFG_POST_MSG("speed     : %d", cable.speed);
+    CFG_POST_MSG("transport : %d", cable.transport);
+    CFG_POST_MSG("channel   : %d", cable.channel);
+  }
+
+  CFG_POST_MSG("-- ListDevices(cables, outputMsg) --")
+  if (cables.size() > 0) {
+    ret = ListDevices(cables[0], devices);
+    if (ret != NoError) {
+      CFG_POST_MSG("Error code = %d. %s", ret, GetErrorMessage(ret).c_str());
+      return ret;
+    }
     CFG_POST_MSG("### ListDevices API testing ###");
     for (auto& device : devices) {
       CFG_POST_MSG("-------------------------");
       CFG_POST_MSG("Device name: %s", device.name.c_str());
-      CFG_POST_MSG("Device index: %s", std::to_string(device.index).c_str());
-      CFG_POST_MSG("Device jtagId: %s", device.jtagId.c_str());
-      CFG_POST_MSG("Device mask: %s", std::to_string(device.mask).c_str());
-      CFG_POST_MSG("Device irlength: %s",
-                   std::to_string(device.irlength).c_str());
-      CFG_POST_MSG("Device flashSize: %s",
-                   std::to_string(device.flashSize).c_str());
+      CFG_POST_MSG("Device index: %d", device.index);
+      CFG_POST_MSG("Device jtagId: 0x%08x", device.tapInfo.idCode);
+      CFG_POST_MSG("Device mask: 0x%08x", device.tapInfo.irMask);
+      CFG_POST_MSG("Device irlength: %d", device.tapInfo.irLen);
+      CFG_POST_MSG("Device flashSize: %s bits",
+                   formatMemorySize(device.flashSize).c_str());
+
+      CFG_POST_MSG("-------------------------");
     }
-    CFG_POST_MSG("-------------------------");
-    CFG_POST_MSG(outputMsg.c_str());
   }
-
-  Device device = devices[0];
-  CfgStatus status;
-  if (GetFpgaStatus(device, status)) {
+  if (devices.size() > 0) {
+    CfgStatus status;
+    ret = GetFpgaStatus(cables[0], devices[0], status);
+    if (ret != NoError) {
+      CFG_POST_MSG("Error code = %d. %s", ret, GetErrorMessage(ret).c_str());
+      return ret;
+    }
     CFG_POST_MSG("### GetFpgaStatus API testing ###");
     CFG_POST_MSG("Device cfgDone: %s", std::to_string(status.cfgDone).c_str());
     CFG_POST_MSG("Device cfgError: %s",
                  std::to_string(status.cfgError).c_str());
   }
 
-  OutputCallback callback = [](std::string msg) {
-    CFG_POST_MSG("### OutputCallback API testing ###");
-    CFG_POST_MSG("OutputCallback msg: %s", msg.c_str());
-  };
-  ProgressCallback progress = [](double percent) {
-    CFG_POST_MSG("### ProgressCallback API testing ###");
-    CFG_POST_MSG("ProgressCallback percent: %s",
-                 std::to_string(percent).c_str());
-  };
+  int userInput = 1;
   std::atomic<bool> stop = false;
-  ProgramFpga(device, "test.bit", "config.cfg", nullptr, callback, progress,
-              stop);
+  std::string bitfile =
```

Comment:
before we continue any operation, shouldn't we check device.size(). 

It is possible that at line 90, when we call ListDevices(), the ret is good but no device is detected right? If yes, without checking devices.size() and just call devices[0] at line 130 and 138 will cause invalid memory access.

Reply:
Yes. I agree. I will fixed that.